cloud/aws/deploy/website.go

```diff
@@ -0,0 +1,396 @@
+	deploymentspb "github.com/nitrictech/nitric/core/pkg/proto/deployments/v1"
+	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/cloudfront"
+	"github.com/pulumi/pulumi-aws/sdk/v5/go/aws/s3"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	awscloudfront "github.com/aws/aws-sdk-go-v2/service/cloudfront"
+	awscloudfronttypes "github.com/aws/aws-sdk-go-v2/service/cloudfront/types"
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+// createBucket - creates a new S3 bucket in AWS and tags it.
+func (a *NitricAwsPulumiProvider) createWebsiteBucket(ctx *pulumi.Context) error {
+	var err error
+
+	name := "website-bucket"
+	tags := common.Tags(a.StackId, name, resources.Bucket)
+
+	a.publicWebsiteBucket, err = s3.NewBucket(ctx, name, &s3.BucketArgs{
+		Tags: pulumi.ToStringMap(tags),
+	})
+	if err != nil {
+		return err
+	}
+
+	a.Buckets[name] = a.publicWebsiteBucket
+
+	return nil
+}
+
+// Website - Implements the Website deployment method for the AWS provider
+func (a *NitricAwsPulumiProvider) Website(ctx *pulumi.Context, parent pulumi.Resource, name string, config *deploymentspb.Website) error {
+	opts := []pulumi.ResourceOption{pulumi.Parent(parent), pulumi.DependsOn([]pulumi.Resource{a.publicWebsiteBucket})}
+
+	cleanedPath := filepath.ToSlash(filepath.Clean(config.OutputDirectory))
+
+	if config.BasePath == "/" {
+		a.websiteIndexDocument = config.IndexDocument
+		a.websiteErrorDocument = config.ErrorDocument
+	}
+
+	// Enumerate the public directory in pwd and upload all files to the public bucket
+	// This will be the source for our cloudfront distribution
+	err := filepath.WalkDir(cleanedPath, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		// Skip directories
+		if d.IsDir() {
+			return nil
+		}
+		// Determine the content type based on the file extension
+		contentType := mime.TypeByExtension(filepath.Ext(path))
+		if contentType == "" {
+			contentType = "application/octet-stream"
+		}
+
+		// Generate the object key to include the folder structure
+		var objectKey string
+
+		filePath := path[len(cleanedPath):]
+
+		arn := filepath.ToSlash(filepath.Join(name, filePath))
+
+		// If the base path is not the root, include it in the object key
+		if config.BasePath == "/" {
+			objectKey = filepath.ToSlash(filePath)
+		} else {
+			objectKey = filepath.ToSlash(filepath.Join(config.BasePath, filePath))
+		}
+
+		output := a.publicWebsiteBucket.Bucket.ApplyT(func(bucket string) (pulumi.StringOutput, error) {
+			existingObj, err := s3.GetObject(ctx, &s3.GetObjectArgs{
+				Bucket: bucket,
+				Key:    objectKey,
+			})
+			if err != nil {
+				// TODO: Could find a better way to check if the object doesn't exist
+				if !strings.Contains(err.Error(), "couldn't find resource") {
+					// Propagate other errors
+					return pulumi.String("").ToStringOutput(), err
+				}
+			}
+
+			obj, err := s3.NewBucketObject(ctx, arn, &s3.BucketObjectArgs{
```

Comment:
Could a pulumi resource be used here instead? https://www.pulumi.com/registry/packages/aws/api-docs/s3/bucketobject/

Reply:
> Could a pulumi resource be used here instead? https://www.pulumi.com/registry/packages/aws/api-docs/s3/bucketobject/

When using the pulumi resource, If a bucket object is deleted or does not exist yet it will fail the entire deployment

Reply:
@davemooreuws I can understand the object being deleted causing issues, but do you have a reproduction for it failing when the object doesn't exist that one sounds like a major bug and there are many pulumi example that just wouldn't work if that was the case.

I think for the deletion case we can accept as users should not be manually removing files from these buckets, lest they break their own site.

Reply:
@tjholm I think I can switch to the resource for aws, the users can just refresh if it gets out of sync.
I believe it was azure that had the read error issues using pulumi, had to use the sdk for that.

Reply:
Did a refactor to use the s3.HeadObject method to get the Etag. This is more performant and we can correctly check the errors. The pulumi read resource methods don't work for this use case (at the moment).